core/libp2p/protocol_muxer/multiselect/multiselect.hpp

```diff
@@ -38,125 +44,85 @@ namespace libp2p::protocol_muxer {
 
     ~Multiselect() override = default;
 
-    outcome::result<peer::Protocol> selectOneOf(
-        gsl::span<const peer::Protocol> supported_protocols,
-        std::shared_ptr<basic::ReadWriteCloser> connection,
-        bool is_initiator) const override;
-
-    enum class MultiselectError {
-      PROTOCOLS_LIST_EMPTY = 1,
-      NEGOTIATION_FAILED,
-      INTERNAL_ERROR
-    };
+    void selectOneOf(gsl::span<const peer::Protocol> protocols,
+                     std::shared_ptr<basic::ReadWriter> connection,
+                     bool is_initiator, ProtocolHandlerFunc cb) override;
 
    private:
-    enum class Status {
-      NOTHING_SENT,
-      OPENING_SENT,
-      PROTOCOL_SENT,
-      PROTOCOLS_SENT,
-      LS_SENT,
-      NA_SENT,
-      NEGOTIATION_SUCCESS,
-      NEGOTIATION_FAIL
-    };
-
     /**
      * Negotiate about a protocol
      * @param connection to be negotiated over
      * @param round, about which protocol the negotiation is to take place
      * @return chosen protocol in case of success, error otherwise
      */
-    outcome::result<peer::Protocol> negotiate(
-        const std::shared_ptr<basic::ReadWriteCloser> &connection,
-        gsl::span<const peer::Protocol> protocols, Status initial_status) const;
+    void negotiate(const std::shared_ptr<basic::ReadWriter> &connection,
+                   gsl::span<const peer::Protocol> protocols, bool is_initiator,
+                   ProtocolHandlerFunc handler);
 
     /**
-     * Finish a negotiation process
-     * @param status, in which the negotiation ended
-     * @param protocol, which was (or not) chosen
-     * @return chosen protocol in case of success, error otherwise
+     * Triggered, when error happens during the negotiation round
+     * @param connection_state - state of the connection
+     * @param ec - error, which happened
      */
-    outcome::result<peer::Protocol> finalizeNegotiation(
-        Status status, const peer::Protocol &protocol) const;
+    void negotiationRoundFailed(
+        const std::shared_ptr<ConnectionState> &connection_state,
+        const std::error_code &ec);
 
-    /**
-     * Handle a message, signalizing about start of the negotiation
-     * @param connection, over which the message came
-     * @param status of the negotiation process
-     * @return status after message handling
-     */
-    outcome::result<Status> handleOpeningMsg(
-        const std::shared_ptr<basic::ReadWriteCloser> &connection,
-        Status status) const;
+    void onWriteCompleted(
```

Comment:
Comments to all methods, please

Reply:
comments like `invoked when write is completed` ? Don't think it makes a lot of sense